nesta/core/orms/orm_utils.py

```diff
@@ -458,23 +609,27 @@ def insert_data(db_env, section, database, Base,
         :obj:`list` of :obj:`dict` data found already existing in the database (optional)
         :obj:`list` of :obj:`dict` data which could not be imported (optional)
     """
-
-    response = filter_out_duplicates(db_env=db_env,
-                                     section=section,
-                                     database=database,
-                                     Base=Base,
-                                     _class=_class,
-                                     data=data,
-                                     low_memory=low_memory)
+    filter_function = merge_duplicates if merge_non_null else filter_out_duplicates
+    response = filter_function(db_env=db_env,
+                               section=section,
+                               database=database,
+                               Base=Base,
+                               _class=_class,
+                               data=data,
+                               low_memory=low_memory)
     objs, existing_objs, failed_objs = response
-    # save and commit
-    engine = get_mysql_engine(db_env, section, database)
-    try_until_allowed(Base.metadata.create_all, engine)
-    Session = try_until_allowed(sessionmaker, engine)
-    session = try_until_allowed(Session)
-    session.bulk_save_objects(objs)
+    # Open a transaction
+    session = get_session(db_env, section, database, Base)
+    if merge_non_null:
+        session.execute(existing_objs)  # Drop existing objs if merging
+    # Insert data in chunks
+    for chunk in split_batches(objs, insert_chunksize):
```

Comment:
note: this is where `bulk_save_objects` (which takes an ORM instance as an argument) is replace by the SQL alchemy native `insert().values()`, which turns out to be significantly faster